File: fast-sync-core/src/core/implementation/fast_sync.ts
import { ISyncableRepository } from "../abstraction/data/ISyncable_Repository";
import { ISyncableObject } from "../abstraction/metadata/ISyncable_object";
import { SyncConfiguration } from "./service/sync_config";
import { SyncVersionManager } from "./service/sync_version_manager";
import { IConflictsHandler } from "../abstraction/service/IConflicts_handler";
import { SyncZoneRestrictionEnum } from "../abstraction/models/Sync_zone_restriction";
import { ISyncManager } from "../abstraction/service/ISync_manager";

export class FastSync {
    private static instance: FastSync;
    private syncConfiguration: SyncConfiguration;
  
    private constructor() {}
  
    public static getInstance(): FastSync;
    public static getInstance(syncConfiguration: SyncConfiguration): FastSync;
  
    public static getInstance(syncConfiguration?: SyncConfiguration): FastSync {
      if (!FastSync.instance) {
        FastSync.instance = new FastSync();
      }
      if (syncConfiguration) {
        FastSync.setSyncConfiguration(syncConfiguration);
      }
      return FastSync.instance;
    }

    public getSyncVersionManager(): SyncVersionManager{
        return FastSync.instance.syncConfiguration.syncVersionManager;
    }
  
    private static setSyncConfiguration(syncConfiguration: SyncConfiguration) {
      FastSync.instance.syncConfiguration = syncConfiguration;
    }     

    public async setSyncalbeObject<T extends ISyncableObject>(entityType: string, repository: ISyncableRepository<T>, conflictsHandler?: IConflictsHandler,  syncZoneRestriction: SyncZoneRestrictionEnum = SyncZoneRestrictionEnum.global ) {
      let syncConfiguration= FastSync.instance.syncConfiguration;
      await syncConfiguration.setSyncalbeObject(entityType,repository,syncZoneRestriction, conflictsHandler);  
      FastSync.setSyncConfiguration(syncConfiguration);
    }

    public getObjectConflictsHandler(type:string): IConflictsHandler {
      return FastSync.instance.syncConfiguration.getObjectConflictsHandler(type);
    }

    public getObjectRepository<T extends ISyncableObject>(type: string): ISyncableRepository<T> {
      return FastSync.instance.syncConfiguration.getObjectRepository(type);
    }

    getSyncZoneConfiguration(type:string): SyncZoneRestrictionEnum{
      return FastSync.instance.syncConfiguration.getSyncZoneConfiguration(type);
    }

    getSyncManager(): ISyncManager{
      return FastSync.instance.syncConfiguration.getSyncManager();
    }

  }
  

File: fast_sync_dotnet_service/fast-sync-core/implementation/FastSync.cs
using fast_sync_core.abstraction.data;
using fast_sync_core.implementation.data;
using fast_sync_core.implementation.metadata;

namespace fast_sync_core.implementation
{
    public class FastSync
    {
        private static FastSync _instance = new FastSync();
        private SyncConfiguration? _syncConfiguration;

        private FastSync() { }

        public static FastSync GetInstance(SyncConfiguration? syncConfiguration = null)
        {
            if (_instance == null)
            {
                _instance = new FastSync();
            }

            if (syncConfiguration != null)
            {
                SetSyncConfiguration(syncConfiguration);
            }

            return _instance;
        }

        public static SyncVersionManager GetSyncVersionManager()
        {
            return GetSyncConfiguration().SyncVersionManager;
        }

        private static void SetSyncConfiguration(SyncConfiguration syncConfiguration)
        {
            _instance._syncConfiguration = syncConfiguration;
        }

        public static void SetSyncableObject<T>(Type type, ISyncableRepository<T> repository, IConflictsHandler conflictsHandler, SyncZoneRestrictionEnum syncZoneRestriction = SyncZoneRestrictionEnum.Global) where T : SyncableObject
        {
            var syncConfiguration = GetSyncConfiguration();
            syncConfiguration.SetSyncableObject(type, repository, syncZoneRestriction, conflictsHandler);
            SetSyncConfiguration(syncConfiguration);
        }

        public static Type getObjectType(string entityType)
        {
            return GetSyncConfiguration().getObjectType(entityType);
        }

        public IConflictsHandler GetObjectConflictsHandler(string type)
        {
            return GetSyncConfiguration().GetObjectConflictsHandler(type);
        }

        public static object GetObjectRepository<T>(string type) where T : SyncableObject
        {
            return GetSyncConfiguration().GetObjectRepository<T>(type);
        }

        public SyncZoneRestrictionEnum GetSyncZoneConfiguration(string type)
        {
            return GetSyncConfiguration().GetSyncZoneConfiguration(type);
        }

        public static ISyncManager GetSyncManager()
        {
            return GetSyncConfiguration().GetSyncManager();
        }

        public static SyncConfiguration GetSyncConfiguration()
        {
            if(_instance._syncConfiguration != null)
            {
                return _instance._syncConfiguration;
            }
            else
            {
                throw new InvalidOperationException("Please init the SyncConfiguration, FastSync.GetInstance(configuration)");
            }
        }
    }
}
